keps/sig-storage/20200609-csi-service-account-token.md

```diff
@@ -0,0 +1,144 @@
+  for AWS credentials.
+
+## Proposal
+
+Kubelet should inject audience-bounded service account tokens requested by the
+CSI driver from the `TokenRequest` API in `VolumeContext` to `NodePublishVolume`
+call.
+
+### CSIDriver Object
+
+```go
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+    ... // existing fields
+
+    RequiresRemount *bool
+    ServiceAccountTokenAudiences []string
+    ServiceAccountTokenExpirationSeconds *int64
+}
+```
+
+These three fields are all optional:
+
+**`ServiceAccountTokenAudiences`**: will be set in `TokenRequestSpec`. All the
+audiences will be validated to not be empty (i.e. kube apiserver). SP of the CSI
+driver is supposed to send a `TokenReview` with at least one of the audiences
+specified.
+
+**`ServiceAccountTokenExpirationSeconds`**: will be set in `TokenRequestSpec`.
+The issued token may have a different duration, so the `ExpirationTimestamp` in
+`TokenRequestStatus` will be passed to CSI driver.
+
+**`RequiresRemount`**: is optional, which should be only set when the mounted
+volumes by the CRI driver have TTL and require re-validation on the token.
+
+The token will be bounded to the pod that the CSI driver is mounting volumes for
+and will be set in `VolumeContext`:
+
+```go
+"csi.storage.k8s.io/serviceAccount.token": token
+"csi.storage.k8s.io/serviceAccount.token.expiry": expirationTimestamp
+```
+
+With the new fields:
+
+1.  The issued token for CSI driver cannot authenticated to kube apiserver.
+    Storage provider is expected to call TokenReview api where audience is
+    specified to validate the token.
+2.  If storage provider requires periodic authentication, `RequiresRemount` can
+    be used to trigger a sequence of `NodePublishVolume` calls. Each call will
+    have valid token in the `VolumeContext`.
+
+### Example Workflow
+
+Take the Vault provider for secret store CSI driver as an example:
+
+1.  Create `CSIDriver` object with `ServiceAccountTokenAudiences=['vault']` and
+    `RequiresRemount=true`
+2.  When the volume manager of kubelet sees a new volume, the pod object in
+    `mountedPods` will have `requiresRemound=true` after `MarkRemountRequired`
+    is called. `MarkRemountRequired` will call into `RequiresRemount` of the
+    in-tree csi plugin to fetch the `CSIDriver` object.
+3.  Before `NodePublishVolume` call, kubelet will request token from
+    `TokenRequest` api with `audiences=['vault']`.
+4.  The token will be specified in `VolumeContext` to `NodePublishVolume` call.
+5.  Every 0.1 second, the reconciler component of volume manager will remount
+    the volume in case the vault secrets expire and re-login is required.
+
+## Alternatives
+
+1.  Instead of fetching tokens in kubelet, CSI drivers will be granted
+    permission to `TokenRequest` api. This will require non-trivial admission
+    plugin to do necessary validation and every csi driver needs to reimplement
+    the same functionality.
+
+## Implementation History
```

Comment:
We also need to fill out the test plan, graduation criteria, and production readiness sections before we can move the kep to implementable.

Reply:
regarding criteria and production readiness, seems they need FeatureGate. i am not sure we need feature gate for this one as it doesn't change existing behavior.

Reply:
I think any API change we need to make would require going through the alpha/beta/ga feature gates so that we have the ability to make changes after testing and receiving feedback.

Reply:
done